zebra-network/src/protocol/external/addr/v2.rs

```diff
@@ -0,0 +1,213 @@
+//! Zcash `addrv2` message node address serialization.
+//!
+//! Zebra parses received IPv4 and IPv6 addresses in the [`AddrV2`] format.
+//! But it ignores all other address types.
+//!
+//! Zebra never sends `addrv2` messages, because peers still accept `addr` (v1) messages.
+
+use std::{
+    convert::TryInto,
+    io::Read,
+    net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr},
+};
+
+use byteorder::{BigEndian, ReadBytesExt};
+
+use zebra_chain::serialization::{
+    CompactSize64, DateTime32, SerializationError, TrustedPreallocate, ZcashDeserialize,
+    ZcashDeserializeInto,
+};
+
+use crate::{
+    meta_addr::MetaAddr,
+    protocol::external::{types::PeerServices, MAX_PROTOCOL_MESSAGE_LEN},
+};
+
+#[cfg(any(test, feature = "proptest-impl"))]
+use proptest_derive::Arbitrary;
+
+/// The maximum permitted size of the `addr` field in `addrv2` messages.
+///
+/// > Field addr has a variable length, with a maximum of 512 bytes (4096 bits).
+/// > Clients MUST reject messages with a longer addr field, irrespective of the network ID.
+///
+/// https://zips.z.cash/zip-0155#specification
+pub const MAX_ADDR_V2_ADDR_SIZE: usize = 512;
+
+/// The size of [`Ipv4Addr`]s in `addrv2` messages.
+///
+/// https://zips.z.cash/zip-0155#specification
+pub const ADDR_V2_IPV4_ADDR_SIZE: usize = 4;
+
+/// The size of [`Ipv6Addr`]s in `addrv2` messages.
+///
+/// https://zips.z.cash/zip-0155#specification
+pub const ADDR_V2_IPV6_ADDR_SIZE: usize = 16;
+
+/// A version 2 node address, its advertised services, and last-seen time.
```

Comment:
Nit: Not sure if this is useful, but I think there might be a tiny ambiguity in `a version 2 node address`:

1. An address of a version 2 node
2. A version 2 address of a node

Not sure if it's worth to reword this. If there's an easy way to change this, then maybe it's a good idea. The risk of miss-understanding is pretty low due to the context, so this could be left as is.

```suggestion
/// A version 2 address of a node, its advertised services, and last-seen time.
```

Reply:
I agree it's confusing, what do you think of af76e24?